src/index.tsx

```diff
@@ -140,22 +142,35 @@ const VisGraph = forwardRef<
     }
     const { added, removed, updated } = diff(prevNodes.current, graph.nodes);
 
-    nodes.remove(removed);
-    nodes.add(added);
-    nodes.update(updated);
-    prevNodes.current = graph.nodes;
+    if (removed.length) {
+      nodes.remove(removed);
+    }
+    // Shallow clone nodes to ensure props aren't mutated!
+    if (added.length) {
+      nodes.add(shallowClone(added));
+    }
+    if (updated.length) {
+      nodes.update(shallowClone(updated));
+    }
+    prevNodes.current = nodes.get();
```

Comment:
I need to keep this as it was (same with prevEdges) which will work better with the isEqual calls. And I can just run `nodes.get` through the `diff` function.

Though I'm thinking it might make sense to remove the isEqual calls anyway... Since theoretically as long as the user doesn't generate the nodes/edges on the fly, it'll have changes, and if they do, as long as they have IDs, we'll more or less be doing the same thing. If they don't have IDs, then we'd be removing/adding them every time